core/arch/x86_code_test.c

```diff
@@ -0,0 +1,351 @@
+#endif
+
+#if defined(STANDALONE_UNIT_TEST)
+
+#    define CONST_BYTE 0x1f
+#    define TEST_STACK_SIZE 4096
+/* Align stack to 16 bytes: sufficient for all current architectures. */
+byte ALIGN_VAR(16) test_stack[TEST_STACK_SIZE];
+static dcontext_t *static_dc;
+
+static void
+check_var(byte *var)
+{
+    EXPECT(*var, CONST_BYTE);
+}
+
+static void (*check_var_ptr)(byte *) = check_var;
+
+static void
+test_func(dcontext_t *dcontext)
+{
+    /* i#1577: we want to read the stack without bothering with a separate
+     * assembly routine and without getting an uninit var warning from the
+     * compiler.  We go through a separate function and avoid compiler analysis
+     * of that function via an indirect call.
+     */
+    byte var;
+    check_var_ptr(&var);
+    EXPECT((ptr_uint_t)dcontext, (ptr_uint_t)static_dc);
+    return;
+}
+
+static void
+test_call_switch_stack(dcontext_t *dc)
+{
+    byte *stack_ptr = test_stack + TEST_STACK_SIZE;
+    static_dc = dc;
+    print_file(STDERR, "testing asm call_switch_stack\n");
+    memset(test_stack, CONST_BYTE, sizeof(test_stack));
+    call_switch_stack(dc, stack_ptr, (void (*)(void *))test_func, NULL,
+                      true /* should return */);
+}
+
+static void
+test_cpuid()
+{
+#    ifdef X86
+    int cpuid_res[4] = { 0 };
+#    endif
+    print_file(STDERR, "testing asm cpuid\n");
+    EXPECT(cpuid_supported(), IF_X86_ELSE(true, false));
+#    ifdef X86
+    our_cpuid(cpuid_res, 0, 0); /* get vendor id */
+    /* cpuid_res[1..3] stores vendor info like "GenuineIntel" or "AuthenticAMD" for X86 */
+    EXPECT_NE(cpuid_res[1], 0);
+    EXPECT_NE(cpuid_res[2], 0);
+    EXPECT_NE(cpuid_res[3], 0);
+#    endif
+}
+
+#    ifdef __AVX__
+
+static void
+unit_test_get_ymm_caller_saved()
+{
+    /* XXX i#1312: Once get_ymm_caller_saved(byte* buf) changes to reflect a dr_zmm_t type
+     * of the SIMD field in DynamoRIO's mcontext, this needs to become dr_zmm_t.
+     */
+    dr_ymm_t ref_buffer[MCXT_NUM_SIMD_SLOTS];
+    dr_ymm_t get_buffer[MCXT_NUM_SIMD_SLOTS];
+    ASSERT(sizeof(dr_ymm_t) == YMM_REG_SIZE);
+    uint base = 0x78abcdef;
+
+    register __m256 ymm0 asm("ymm0");
+    register __m256 ymm1 asm("ymm1");
+    register __m256 ymm2 asm("ymm2");
+    register __m256 ymm3 asm("ymm3");
+    register __m256 ymm4 asm("ymm4");
+    register __m256 ymm5 asm("ymm5");
+    register __m256 ymm6 asm("ymm6");
+    register __m256 ymm7 asm("ymm7");
+#        ifdef X64
+    register __m256 ymm8 asm("ymm8");
+    register __m256 ymm9 asm("ymm9");
+    register __m256 ymm10 asm("ymm10");
+    register __m256 ymm11 asm("ymm11");
+    register __m256 ymm12 asm("ymm12");
+    register __m256 ymm13 asm("ymm13");
+    register __m256 ymm14 asm("ymm14");
+    register __m256 ymm15 asm("ymm15");
+#        endif
+
+    for (int regno = 0; regno < proc_num_simd_registers(); ++regno) {
+        for (int dword = 0; dword < sizeof(dr_ymm_t) / sizeof(uint); ++dword) {
+            ref_buffer[regno].u32[dword] = 0;
```

Comment:
Why set this to 0 if it's going to be clobbered on line 146?  Seems like this line should be removed.

Reply:
Yes I can also remove the second loop, sure. Done.